new module: AIX Volume Group creating, resizing, removing

kairoaraujo opened the pull request:
##### SUMMARY
It is a module to manage Volume Groups to AIX Logical Volume
Manager. With this module is able to create, resize and
remove a Volume Group.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
aix_lvg

##### ANSIBLE VERSION
```
ansible 2.5.0 (aix_lvg bfeb840b21) last updated 2017/09/14 22:05:36 (GMT +200)
  config file = None
  configured module search path = ['/Users/kairoaraujo/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/kairoaraujo/Dev/aix_lvg/ansible/lib/ansible
  executable location = /Users/kairoaraujo/Dev/aix_lvg/ansible/bin/ansible
  python version = 3.6.2 (v3.6.2:5fd33b5926, Jul 16 2017, 20:11:06) [GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
```


##### ADDITIONAL INFORMATION
None

kairoaraujo committed:
new module: AIX Volume Group creating, resizing, removing

It is a module to manage Volume Groups to AIX Logical Volume
Manager. With this module is able to create, reisize and
remove a Volume Group.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+short_description: Configure LVM volume groups for AIX
+description:
+  - This module creates, removes or resize volume groups on AIX LVM.
+version_added: 2.5
```
The value should be quoted.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
```
Please format this like this:

```python
ANSIBLE_METADATA = {
    'metadata_version': '1.1',
    'status': ['preview'],
    'supported_by': 'community'
}
```

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+options:
+  vg:
+    description:
+    - Volume Group name
```
Please indent the array items by two characters like you did for example above in the case of the `description`. The same in all cases bellow.

Also make the `Group` lowercase to unify it across all documentation. And add period at the end of the sentence.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+  vg_type:
+    description:
+    - Volume Group type.
+    choices: ["normal", "big", "scalable"]
```
The values in the array should not be quoted.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+  state:
+    description:
+    - Control if the volume group exists.
+    choices: [ "present", "absent" ]
```
The values in the array should not be quoted.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+  force:
+    description:
+    - Forces volume group creation
+    choices: ['yes', 'no']
```
This should be quoted but please use double quotes.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+    description:
+    - Forces volume group creation
+    choices: ['yes', 'no']
+    default: no
```
The value should be quoted.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+    default: no
+notes:
+  - AIX will permit remove VG only if all LV/Filesystems aren't busy
+  - module does not modify PP size for already present volume group
```
Capital letter at the beginning and period at the end.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+    choices: ['yes', 'no']
+    default: no
+notes:
+  - AIX will permit remove VG only if all LV/Filesystems aren't busy
```
Missing period at the end.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+EXAMPLES = '''
+# Create a volume group datavg
+aix_lvg:
+    vg: datavg
```
Standard YAM indentation is two characters. Please change it to maintain the same indentation like in the `DOCUMENTATION` block. The same for all examples.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+'''
+
+EXAMPLES = '''
+# Create a volume group datavg
```
Please make the comment as the task name:

```yaml
- name: Create a volume group datavg
  aix_lvg:
    pp_size: 128
    vg_type: scalable
    state: present
```

The same for all examples.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+aix_lvg:
+    vg: datavg
+    pp_size: 128
+    vg_type: 'scalable'
```
The value should not be quoted.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+
+RETURN = '''
+changed:
+    description: Return changed for aix_lvg actions as true or false.
```
Standard YAM indentation is two characters. Please change it to maintain the same indentation like in the `DOCUMENTATION` block.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+            vg_type=dict(choices=["normal", "big", "scalable"],
+                         default="normal"),
+            state=dict(choices=["absent", "present"], default="present"),
+            force=dict(type=bool, default=False)
```
The `type` value should be quoted:

```python
force=dict(type="bool", default=False)
```

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+    if state == 'present':
+
+        if not pvs:
+            module.fail_json(msg="pvs is required to state 'present' ")
```
Please remove the space at the end of the message.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+        argument_spec=dict(
+            vg=dict(required=True),
+            pvs=dict(type="list", default=None),
+            pp_size=dict(choices=["1", "2", "4", "8", "16", "32", "64", "128",
```
Shouldn't this be `type="int"`?

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+        for line in current_pvs.splitlines()[2:]:
+                pvs_to_remove.append(line.split()[0])
+
+        reduce_msg = ("Volume group %s removed" % vg)
```
You can remove the brackets here.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+
+def _validate_pv(module, pvname):
+    """
+    function to validate if the Physical Volume is not already in use by
```
Capital letter at the beginning and period at the end of the sentence.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+    required: false
+  vg_type:
+    description:
+    - Volume Group type.
```
Make the `Group` lowercase to unify it across all documentation.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+  pvs:
+    description:
+    - List of comma-separated devices to use as physical devices in this volume
+      group. Required when creating or extending ('present' state) the volume
```
`present` is a value of a parameter so it should be decorated like this: `C(present)`. The same for the `absent` bellow.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+
+DOCUMENTATION = '''
+---
+author: "Kairo Araujo (@kairoaraujo)"
```
The value doesn't have to be quoted.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,273 @@
+    required: false
+  force:
+    description:
+    - Forces volume group creation
```
Missing period at the end of sentence.

molekuul commented:
Hi,
Nice to see someone is working on AIX, I thought I was the onlyone :-)

I have some remarks on the module:
Ansible modules are to set a status of a configuration, if you want to extend a volumegroup with a disk that's already in this volumegroup, the module should not fail, but should return changed: False, nothing changed. 

When a disk is in use by a volumegroup, you exit the module with a fail,
but when the disk is in use as an ASM disk, you set the return to False, and let the create_extend_vg function handle this, why these different approaches?

If you give a list of disks, the module will fail if even one disk will fail the "usage check" is this wat you want?

Checkmode is a mode in Ansible where you check what will be changed when you run the module with these parameters, without actually doing the action. You didn't incorporate this in the module, however you state that you support checkmode.

There are limits of how many PP's a disk can contain. An AIX system calulates the PP size from the size of the disk if you do not specify this. why not let the system decide the value of the PP-size if not specified in the playbook?

When you are removing a pv from a volumegroup, you do not check if the pv is within this volumegroup, so the command will fail, but the result is correct, the disk is not part of the volumegroup, as you want, you should handle this situation better.

Take into account that a volumegroup must be varyon before taking actions on a volumegroup.

I hope you can do anything with my remarks. 

Kind regards,

Joris Weijters

kairoaraujo commented:
Hi @molekuul 

Yes, as AIX is my main OS I am writing some modules, I have also almost ready to submit modules for filesystem and mksysb. Remarks are always welcome :) 
Please, could you also check out my packaging/os/installp PR #30238 ?

I will go through your remarks and see I'll what I can do.

kairoaraujo committed:
fixed suggestion for standards and PEP8

fixed suggestion on PR according with standards and PEP8

kairoaraujo committed:
Removed blank space in the end of line.

Removed blank space in the end of line.

kairoaraujo commented:
Working on @molekuul remarks.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,275 @@
+  type: boolean
+  version_added: 2.5
+msg:
+    description: Return message regarding the action.
```
Please indent it by 2 spaces only.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,275 @@
+            state=dict(choices=["absent", "present"], default="present"),
+            force=dict(type="bool", default=False)
+        ),
+        supports_check_mode=True,
```
You claim support for the check mode but I don't see any implementation here. Please could you explain?

kairoaraujo replied:
I was confused how to implement it but your question clarified :)
I need to implement it.

dagwieers replied:
This has been resolved.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,275 @@
+    force = module.params['force']
+
+    if state == 'present':
+
```
I would remove this blank line.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,275 @@
+
+        if not pvs:
+            module.fail_json(msg="pvs is required to state 'present'")
+
```
I would remove this blank line.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,275 @@
+                pvs_to_remove.append(line.split()[0])
+
+        reduce_msg = "Volume group %s removed" % vg
+
```
I would remove this blank line.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,275 @@
+
+    if state == 'absent':
+        reduce_vg(module, vg, pvs)
+
```
Here should be two blank lines.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,275 @@
+    lspv_cmd = module.get_bin_path('lspv', True)
+    rc, current_lspv, err = module.run_command('%s ' % lspv_cmd)
+    if rc != 0:
+        module.fail_json(msg="Failed executing lspv command.", rc=rc,
```
I kind of don't like this sort of wrapping. What about to use the following style?

```python
        module.fail_json(
            msg="Failed executing lspv command.", rc=rc, err=err)
```

The same for all similar cases bellow.

kairoaraujo replied:
I can change it if it looks better.

kairoaraujo replied:
@jtyr which one Should I use?

```python
module.fail_json(
            msg='Failed executing lspv command.', rc=rc, err=err)
```
or
```python
module.fail_json(
            msg = 'Failed executing lspv command.', rc = rc, err = err)
```

jtyr replied:
The first one is the only correct one. And you should use double quotes as per the above comment.

kairoaraujo replied:
> Regarding single/double quotes used across the code, there is non-written convention to use double quotes only for text with placeholders (e.g. "%s ") or longer text messages (e.g. msg="Failed executing lspv command."). When a text contains double quotes, to avoid escaping the quotes, single quotes should be used (e.g. my_var = 'This is "my" variable.'). The rest of cases should use single quotes. Please walk through your code and unify the quotes according this convention.

I'm still confused with this non-written convention to use double quotes and I want to have it clear once I want to send more contributions :)

> "The rest of cases should use **single** quotes." 

a. For text with placeholders as ```"%s"``` or ```"{0}".format(variable)``` use double quotes;
b. For long text/sentences as ```msg = "This is an error message."```;
c. Texts with quotes, to avoid escaping single quotes: ```my_var = 'This is "my" variable.'```
d. In case of text with variables, double quote? ```msg = "The VG %s was changed"```

jtyr replied:
That all sounds correct.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,275 @@
+    """
+
+    lspv_cmd = module.get_bin_path('lspv', True)
+    rc, current_lspv, err = module.run_command('%s ' % lspv_cmd)
```
Regarding single/double quotes used across the code, there is non-written convention to use double quotes only for text with placeholders (e.g. `"%s "`) or longer text messages (e.g. `msg="Failed executing lspv command."`). When a text contains double quotes, to avoid escaping the quotes, single quotes should be used (e.g. `my_var = 'This is "my" variable.'`). The rest of cases should use single quotes. Please walk through your code and unify the quotes according this convention.

kairoaraujo replied:
indeed

jtyr replied:
I see did some changes of single quotes to double quotes but you left all double quotes are the were. You should change all other cases than named above to single quotes (e.g. keys in the dict vars - `my_dict['my_key']`).

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,275 @@
+                                           force_mode[force], vg, ' '.join(pvs)
+                                           ))
+        if rc == 0:
+            module.exit_json(changed=True, msg="Volume group %s created" % vg)
```
Please could you change the logic the way that this function returns the changed value and you exit at the end of the `main()` function? The same in for the `reduce_vg()` function.

kairoaraujo commented:
Hi @molekuul 

I would like ask somethings regarding your remarks

> Ansible modules are to set a status of a configuration, if you want to extend a volumegroup with a disk that's already in this volumegroup, the module should not fail, but should return changed: False, nothing changed.

I agree with that point and I'm implementing it.
My doubt is when the disk is already in use by another volume group, it should fail or also return not changed?

> When a disk is in use by a volumegroup, you exit the module with a fail,
but when the disk is in use as an ASM disk, you set the return to False, and let the create_extend_vg function handle this, why these different approaches?
> If you give a list of disks, the module will fail if even one disk will fail the "usage check" is this wat you want?

My approach is wrong, should be the same return for ASM and VG use. IMHO If you give a list of disks and one disk is wrong it should fail and forces the administrator correct the pool and avoid wrong resizes. Anyway, I'm opened to opinions.

> Checkmode is a mode in Ansible where you check what will be changed when you run the module with these parameters, without actually doing the action. You didn't incorporate this in the module, however you state that you support checkmode.

I want to implement it, but I'm quite confused how to do it. 

> There are limits of how many PP's a disk can contain. An AIX system calulates the PP size from the size of the disk if you do not specify this. why not let the system decide the value of the PP-size if not specified in the playbook?

Totally agree, I'm changing it. 

> When you are removing a pv from a volumegroup, you do not check if the pv is within this volumegroup, so the command will fail, but the result is correct, the disk is not part of the volumegroup, as you want, you should handle this situation better.

It is treated by AIX as well, but for sure I do it also.
```
TASK [Reducing datavg with disk from rootvg] **************************************************************************************
fatal: [172.16.100.24]: FAILED! => {"changed": false, "err": "0516-1782 reducevg: Physical volume hdisk0 is not part of volume group datavg.\n0516-884 reducevg: Unable to remove physical volume hdisk0.\n", "failed": true, "msg": "Unable to remove datavg", "rc": 2}
```
We will have also this situation when the volume group still having filesystems mounted or open lv. I think is better leave it up to AIX, no?

> Take into account that a volumegroup must be varyon before taking actions on a volumegroup.

Including this verification on the top.
About this subject, I will also include C(action): varyonv/varyoff it will be useful ;)

Kind regards, 
Kairo Araujo

kairoaraujo committed:
Improved validations, functions, pep8

- fixed pep8 and non-written conventions;
- fixed RETURN doc indentation for msg;
- fixed and improved the physical volume verification;
- pp_size, when not specified, is followed from AIX side;
- included volume group verification in the top;
- included state varyon and varyoff to be easy in playbooks;
- removed check_mode

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+            msg="Failed executing lspv command.", rc=rc, err=err)
+
+    for pv in pvs:
+        # get pv list
```
Comments should start with a capital letter. The same anywhere bellow.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+
+            # check if PV is already in use for the same vg
+            elif vg == lspv_list[pv]:
+                msg = ("Physical volume %s is already used by volume group %s."
```
This should be:

```python
                msg = (
                    "Physical volume %s is already used by volume group %s."
                    % (pv, lspv_list[pv]))
```

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+                return False, msg
+            else:
+                module.fail_json(
+                    msg="Physical volume %s is in use by another volume group "
```
This should be:

```python
                module.fail_json(
                    msg=(
                        "Physical volume %s is in use by another volume group "
                        "%s." % (pv, lspv_list[pv])))
```

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+
+
+def create_extend_vg(module, vg, pvs, pp_size, vg_type, force, vg_validation):
+    """ creates or extend a volume group"""
```
This should be:

```python
    """ Creates or extend a volume group. """
```

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            vg=dict(required=True),
```
Please order the parameters alphabetically.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+  - This module creates, removes or resize volume groups on AIX LVM.
+version_added: "2.5"
+options:
+  vg:
```
Please order the options alphabetically.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+
+
+def reduce_vg(module, vg, pvs, vg_validation):
+
```
You can remove this blank line.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+
+
+def state_vg(module, vg, state, vg_validation):
+
```
You can remove this blank line.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+
+    elif state == 'varyon' or 'varyoff':
+        state_vg(module, vg, state, vg_validation)
+
```
Bere should be two blank lines.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+    description:
+      - Control if the volume group exists and volume group AIX state varyonvg
+        C(varyon) or varyoffvg C(varyoff).
+    choices: [ present, absent, varyon, varyoff ]
```
Remove spaces at the beginning and the end of the square brackets.

jtyr commented:
The Check Mode should be implemented the way that if using `ansible-play -C site.yaml`, the module doesn't perform any real changes, it just reports whether it _would_ perform any change if run without the Check Mode. In your case, you should just check if there already exists VG of that name or parameters and if not, you should not do any changes but just return a flag that a change would be done if running without a Check Mode. Code-wise it would look like this:

```python
def state_vg(...):
    changed = False

    ...

    if state == 'varyon':
        if vg_state is False:
            changed = True

            if not module.check_mode:
                varyonvg_cmd = module.get_bin_path("varyonvg", True)
                rc, varyonvg_out, err = module.run_command(
                    "%s %s" % (varyonvg_cmd, vg))

                if rc != 0:
                    module.failed_json(msg="...", stdout=out, stderr=err)

    ...

    return changed


def main():
    if state == 'present':
        changed = create_extend_vg(...)
    elif state == 'absent':
        changed = reduce_vg(...)
    elif state == 'varyon' or 'varyoff':
￼         changed = state_vg(...)

    module.exit_json(changed=True, state=state)
```

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+        if rc == 0:
+            module.exit_json(changed=True, msg="Volume group %s created." % vg)
+        else:
+            module.exit_json(
```
This should be `fail_json()`.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+        if rc == 0:
+            module.exit_json(changed=True, msg=reduce_msg)
+        else:
+            module.exit_json(
```
This should be `fail_json()`.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+            rc, varyonvg_out, err = module.run_command(
+                "%s %s" % (varyonvg_cmd, vg))
+            if rc != 0:
+                module.exit_json(changed=False, rc=rc, err=err)
```
This should be `fail_json()`.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,365 @@
+            rc, varyonvg_out, err = module.run_command(
+                "%s %s" % (varyonvg_cmd, vg))
+            if rc != 0:
+                module.exit_json(changed=False, msg=err)
```
This should be `fail_json()`.

kairoaraujo committed:
Implemented back module.check_module, pep8, return

Implemented back the module.check_module to permit be used.
Some corrections regarding module.exit_json() and
module.fail_json().
Some pep8 and non-written conventions

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,404 @@
+                        changed=True,
+                        msg="Varyon volume group %s completed." % vg)
+
+            return changed
```
Why do you return something when you are not doing nothing with it in the `main()` function? What about to use `module.exit_json()` only in the `main()` function? The same in other functions.

kairoaraujo replied:
Could I do it as I did in the comment above?

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,404 @@
+            return changed
+
+        else:
+            module.exit_json(
```
I'm sure this can be on one line.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,404 @@
+        if not pvs:
+            module.fail_json(msg='pvs is required to state "present".')
+        else:
+            create_extend_vg(module, vg, pvs, pp_size, vg_type, force,
```
This should follow the same formatting principle like above:

```python
            create_extend_vg(
                module, vg, pvs, pp_size, vg_type, force, vg_validation)
```

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,404 @@
+
+    if state == 'varyon':
+        if vg_state is False:
+            changed = True
```
This should be on the function level to avoid multiple definitions of the same variable on multiple places in the same function.

kairoaraujo replied:
Sorry, maybe I didn't get it. 
I tried to follow the example that you gave me previously and I did a mess.
In this case, this block should be like I did below?

```python
    if state == 'varyon':
        if vg_state is False:
            if not module.check_mode:
                varyonvg_cmd = module.get_bin_path("varyonvg", True)
                rc, varyonvg_out, err = module.run_command(
                    "%s %s" % (varyonvg_cmd, vg))
                if rc != 0:
                    module.fail_json(msg="Varyonvg failed.", rc=rc, err=err)
                else:
                    module.exit_json(
                        changed=True,
                        msg="Varyon volume group %s completed." % vg)
            module.exit_json(changed=True)

        else:
            module.exit_json(changed=False, msg=msg)
```

jtyr replied:
Do not call `exit_json()` from inside of this function. The function should return the `changed` value and you should use `exit_json()` at the end of the `main()` function.

kairoaraujo replied:
Indeed. In that case I will return ```changed```and ```msg``` and use ```exit_json()``` in end of ```main()``` and I will perform it for all other functions.

Regarding ```fail_json()``` Can I keep it into the functions, right? If not, I need to create some standard return like:

```python
# type: fail or exit
# changed Fail or True
return type=fail, changed=False, msg=msg, rc=rc, err=err
```

jtyr replied:
You can keep the `fail_json()` in the functions. Only the `exit_json()` use from the `main()` only.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,404 @@
+
+
+def create_extend_vg(module, vg, pvs, pp_size, vg_type, force, vg_validation):
+    """ Creates or extend a volume group"""
```
This should be:

```python
""" Creates or extend a volume group. """
```

kairoaraujo committed:
removed exit_json() from functions and pep8

- ordered option parameters and pep8 compliance
- removed exit_json() from functions and moved to main()

kairoaraujo committed:
pep8, two spaces before function

pep8, two spaces before function

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,438 @@
+            force=dict(type="bool", default=False),
+            pp_size=dict(type="int", default=None),
+            pvs=dict(type="list", default=None),
+            state=dict(choices=["absent", "present", "varyon", "varyoff"],
```
This should be:

```python
            state=dict(
                choices=["absent", "present", "varyon", "varyoff"],
                default="present"),
```

The same for `vg_type` bellow.

kairoaraujo replied:
Regarding double/single quotes. First, is hard to change some habits. :)

I suppose we are talking regarding ```msg = ''```, right? I was in doubt about it and I checked module mount.py and I got more confused now.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,438 @@
+
+            if rc == 0:
+
+                return changed, reduce_msg
```
I would change this to:

```python
            if rc == 0:
                msg = reduce_msg
```

jtyr reviewed (changes requested):
It looks good now. Just a few comments this time. Also, are you going to address the single/double quotes we discussed previously? Basically everything should be single quotes unless it's formatting string (e.g. `" %s "`) or longer sentence (e.g. "This is my message") or the string contains single quotes (e.g. `"This is 'my' message."`.

kairoaraujo committed:
changed single/double quotes

changed single/double quotes to non-written convention.

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,437 @@
+from ansible.module_utils.basic import AnsibleModule
+
+# Command option parameters.
+force_opt = {
```
I would put these variables into the function which is using them. No reason to have them as global vars if you don't share them across multiple functions.

jtyr reviewed (changes requested):
Otherwise it looks good.

kairoaraujo committed:
removed command options parameters as global vars.

Commands options parameters was moved as global variables
to inside create_vg().

jtyr commented on lib/ansible/modules/system/aix_lvg.py:
```diff
@@ -0,0 +1,437 @@
+        changed = False
+        msg = "Unexpected state"
+
+    module.exit_json(changed=changed, msg=msg)
```
Perhaps you could also return `state=state` here.

kairoaraujo committed:
include state to return.

included the state to return.

jtyr reviewed (approved):
LGTM. `shipit`

kairoaraujo committed:
no-underscore-variable applied to the code

no-underscore-variable was applied to the code

maxamillion commented:
rebuild_merge

maxamillion commented:
@kairoaraujo Apologies for the lag time on this, but if you don't mind fixing up the sanity tests I'm happy to merge.

kairoaraujo commented:
I will take care of it this weekend.

kairoaraujo committed:
Fixed version and doc

Fixed version and documentation sanity

gforster reviewed (commented):
Looks like all that is needed is to change
`  force:
    description:
      - Forces volume group creation.
    choices: [True, False]
    default: "no"
`

to 

`  force:
    description:
      - Forces volume group creation.
    type: bool
    default: "no"
`
If I'm reading the error and the documentation correctly

kairoaraujo committed:
Fixed documentation for E325, E326

Fixed documentation for E325, E326 to force option

maxamillion commented:
needs_revision

gforster commented:
@maxamillion  - what needs revision at this stage?

kairoaraujo commented:
If we have it ok would be nice have it merged as for AIX we have the modules for logical volumes (aix_lvol), filesystems (aix_filesystem). We could say that we will have LVM for AIX well covered by Ansible :)

maxamillion commented:
rebuild_merge

maxamillion commented:
@gforster nothing that I can see, I'm not even sure why I put that on there anymore. From now on I need to provide context because I don't know what past me was thinking :)

kairoaraujo committed:
Fixed E319 for return doc.

kairoaraujo commented:
I will write the integration test for this module as well before merge.

dagwieers committed:
Various improvements and cosmetic changes

dagwieers committed:
Fix whitespace issue
